devices/uart-imx6ull/uart.c

```diff
@@ -0,0 +1,405 @@
+/*
+ * Phoenix-RTOS
+ *
+ * Operating system loader
+ *
+ * i.MX 6ULL Serial driver
+ *
+ * Copyright 2018, 2021 Phoenix Systems
+ * Author: Kamil Amanowicz, Marek Białowas, Hubert Buczynski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <lib/lib.h>
+#include <devices/devs.h>
+
+
+#define UARTS_MAX_CNT 8
+#define BUFFER_SIZE   0x20
+#define MODULE_CLK    20000000
+
+
+typedef struct {
+	volatile u32 *base;
+	unsigned int irq;
+	u16 clk;
+
+	u8 dataRx[BUFFER_SIZE];
+	cbuffer_t cbuffRx;
+
+	u8 dataTx[BUFFER_SIZE];
+	cbuffer_t cbuffTx;
+} uart_t;
+
+
+typedef struct {
+	int ioctl;
+	char val;
+} uart_ioctl_t;
+
+
+enum { urxd = 0, utxd = 16, ucr1 = 32, ucr2, ucr3, ucr4, ufcr, usr1, usr2,
+	uesc, utim, ubir, ubmr, ubrc, onems, uts, umcr };
+
+
+static const uart_ioctl_t uart_mux[UARTS_MAX_CNT][4] = {
+	{ { mux_uart1_cts, 0 }, { mux_uart1_rts, 0 }, { mux_uart1_rx, 0 }, { mux_uart1_tx, 0 } },
+	{ { mux_uart2_cts, 0 }, { mux_uart2_rts, 0 }, { mux_uart2_rx, 0 }, { mux_uart2_tx, 0 } },
+	{ { mux_uart3_cts, 0 }, { mux_uart3_rts, 0 }, { mux_uart3_rx, 0 }, { mux_uart3_tx, 0 } },
+	{ { mux_lcd_hsync, 2 }, { mux_lcd_vsync, 2 }, { mux_uart4_rx, 0 }, { mux_uart4_tx, 0 } },
+	{ { mux_gpio1_09, 8 }, { mux_gpio1_08, 8 }, { mux_uart5_rx, 0 }, { mux_uart5_tx, 0 } },
+	{ { mux_enet1_tx1, 1 }, { mux_enet1_txen, 1 }, { mux_enet2_rx1, 1 }, { mux_enet2_rx0, 1 } },
+	{ { mux_lcd_d6, 1 }, { mux_lcd_d7, 1 }, { mux_lcd_d17, 1 }, { mux_lcd_d16, 1 } },
+	{ { mux_lcd_d4, 1 }, { mux_lcd_d5, 1 }, { mux_lcd_d21, 1 }, { mux_lcd_d20, 1 } },
+};
+
+
+static const uart_ioctl_t uart_isel[UARTS_MAX_CNT][2] = {
+	{ { isel_uart1_rts, 3 }, { isel_uart1_rx, 3 } },
+	{ { isel_uart2_rts, 1 }, { isel_uart2_rx, 1 } },
+	{ { isel_uart3_rts, 1 }, { isel_uart3_rx, 1 } },
+	{ { isel_uart4_rts, 3 }, { isel_uart4_rx, 1 } },
+	{ { isel_uart5_rts, 1 }, { isel_uart5_rx, 7 } },
+	{ { isel_uart6_rts, 3 }, { isel_uart6_rx, 2 } },
+	{ { isel_uart7_rts, 3 }, { isel_uart7_rx, 3 } },
+	{ { isel_uart8_rts, 3 }, { isel_uart8_rx, 3 } },
+};
+
+
+static const struct {
+	volatile u32 *base;
+	unsigned int irq;
+	u32 clk;
+} info[UARTS_MAX_CNT] = {
+	{ (void *)0x02020000, 58, clk_uart1 },
+	{ (void *)0x021E8000, 59, clk_uart2 },
+	{ (void *)0x021EC000, 60, clk_uart3 },
+	{ (void *)0x021F0000, 61, clk_uart4 },
+	{ (void *)0x021F4000, 62, clk_uart5 },
+	{ (void *)0x021FC000, 49, clk_uart6 },
+	{ (void *)0x02018000, 71, clk_uart7 },
+	{ (void *)0x02288000, 72, clk_uart8 }
+};
+
+
+struct {
+	uart_t uarts[UARTS_MAX_CNT];
+	int clkInit;
+} uart_common;
+
+
+static inline void uart_rxData(uart_t *uart)
+{
+	char c;
+
+	/* Keep getting data until fifo is not empty */
+	while (*(uart->base + usr2) & 0x1) {
+		c = *(uart->base + urxd) & 0xff;
```

Comment:
```suggestion
		c = *(uart->base + urxd);
```
Mask is not needed, `c` is type `char`.